internal_worker_base.go

```diff
@@ -121,18 +136,119 @@ func (bw *baseWorker) Start() {
 		return
 	}
 
-	// Launch the routines to do work
-	for i := 0; i < bw.options.routineCount; i++ {
+	for i := 0; i < bw.options.pollerCount; i++ {
 		bw.shutdownWG.Add(1)
-		go bw.execute(i)
+		go bw.runPoller()
 	}
+	bw.shutdownWG.Add(1)
+	go bw.runTaskDispatcher()
 
 	bw.isWorkerStarted = true
 	traceLog(func() {
-		bw.logger.Info("Started Worker", zap.Int("RoutineCount", bw.options.routineCount))
+		bw.logger.Info("Started Worker",
+			zap.Int("PollerCount", bw.options.pollerCount),
+			zap.Int("MaxConcurrentTask", bw.options.maxConcurrentTask),
+			zap.Float32("MaxTaskRps", bw.options.maxTaskRps),
+		)
 	})
 }
 
+func (bw *baseWorker) isShutdown() bool {
+	select {
+	case <-bw.shutdownCh:
+		return true
+	default:
+		return false
+	}
+}
+
+func (bw *baseWorker) runPoller() {
+	defer bw.shutdownWG.Done()
+
+	for {
+		select {
+		case <-bw.shutdownCh:
+			return
+		case <-bw.pollerRequestCh:
+			ch := make(chan struct{})
+			go func(ch chan struct{}) {
+				bw.pollTask()
+				close(ch)
+			}(ch)
+
+			// block until previous poll completed or return immediately when shutdown
+			select {
+			case <-bw.shutdownCh:
+				return
+			case <-ch:
+			}
+		}
+	}
+}
+
+func (bw *baseWorker) runTaskDispatcher() {
+	defer bw.shutdownWG.Done()
+
+	for i := 0; i < bw.options.maxConcurrentTask; i++ {
+		bw.pollerRequestCh <- struct{}{}
+	}
+
+	for {
+		// wait for new task or shutdown
+		select {
+		case <-bw.shutdownCh:
+			return
+		case task := <-bw.taskQueueCh:
+			// block until taskRateLimiter satisfied
+			for !bw.taskRateLimiter.Consume(1, time.Millisecond) {
```

Comment:
can we increase the timeout to 100ms instead of 1ms ? This will avoid spinning in a loop, if we are contending for tokens, while at the same time allows us to shutdown with 100ms.